tests/frontend/test_circuitspecs_chip0.py

```diff
@@ -0,0 +1,300 @@
+import strawberryfields.ops as ops
+from strawberryfields.io import to_program
+from strawberryfields.utils import random_interferometer
+from strawberryfields.circuitspecs.chip0 import Chip0Specs, CircuitSpecs
+
+
+pytestmark = pytest.mark.frontend
+
+np.random.seed(42)
+
+
+class DummyCircuit(CircuitSpecs):
+    """Dummy circuit used to instantiate
+    the abstract base class"""
+    modes = 4
+    remote = False
+    local = True
+    interactive = True
+    primitives = {"S2gate", "MeasureFock", "Rgate", "BSgate"}
+    decompositions = {"Interferometer": {}, "MZgate": {}}
+
+
+class TestChip0Compilation:
+    """Tests for compilation using the Chip0 circuit specification"""
+
+    def test_exact_template(self):
+        """Test compilation works for the exact circuit"""
+        bb = blackbird.loads(Chip0Specs.circuit)
+        bb = bb(
+            squeezing_amplitude_0=0.43,
+            squeezing_amplitude_1=0.65,
+            external_phase_0=0.54,
+            internal_phase_0=-0.23,
+            phi0=1.24,
+            phi1=-0.54,
+            phi2=4.12,
+            phi3=0,
+        )
+
+        expected = to_program(bb)
+        res = expected.compile("chip0")
+
+        for cmd1, cmd2 in zip(res.circuit, expected.circuit):
+            assert cmd1.op.__class__ == cmd2.op.__class__
+            assert [p.x for p in cmd1.op.p] == [p.x for p in cmd2.op.p]
+            assert [i.ind for i in cmd1.reg] == [i.ind for i in cmd2.reg]
+
+    def test_not_all_modes_measured(self):
+        """Test exceptions raised if not all modes are measured"""
+        prog = sf.Program(4)
+        U = random_interferometer(2)
+
+        with prog.context as q:
+            ops.S2gate(0.5) | (q[0], q[2])
+            ops.S2gate(0.5) | (q[1], q[3])
+            ops.Interferometer(U) | (q[0], q[1])
+            ops.Interferometer(U) | (q[2], q[3])
+            ops.MeasureFock() | (q[0], q[1])
+
+        with pytest.raises(CircuitError, match="All modes must be measured"):
+            res = prog.compile("chip0")
+
+    def test_no_s2gates(self):
+        """Test exceptions raised if no S2gates are present"""
+        prog = sf.Program(4)
+        U = random_interferometer(2)
+
+        with prog.context as q:
+            ops.Interferometer(U) | (q[0], q[1])
+            ops.Interferometer(U) | (q[2], q[3])
+            ops.MeasureFock() | q
+
+        with pytest.raises(CircuitError, match="must start with two S2gates"):
+            res = prog.compile("chip0")
+
+    def test_incorrect_s2gates(self):
+        """Test exceptions raised if S2gates are on incorrect modes"""
+        prog = sf.Program(4)
+        U = random_interferometer(2)
+
+        with prog.context as q:
+            ops.S2gate(0.5) | (q[0], q[2])
+            ops.Interferometer(U) | (q[0], q[1])
+            ops.Interferometer(U) | (q[2], q[3])
+            ops.MeasureFock() | q
+
+        with pytest.raises(CircuitError, match="S2gates placed on the incorrect modes"):

```

Comment:
> It would be more clear if the error message said something like "S2gates do not appear on the correct modes"

Will make this change.

> 2nd question: does the particular mode pair matter, so long as it is consistent with the rest of the circuit topologically?

This is the reason why I am not doing much manual checking for the S2gates, and instead I'm just letting the final topology check catch any errors.

I was actually a bit torn with what to do here - we could go even further, and check which S2gates are correct (for more descriptive error messages), and add in identity S2gates if none are provided. But I decided this might be over-complicating it --- including the S2gates in the right place is much easier to do manually than the interferometer.